mirrord/cli/src/port_forward.rs

```diff
@@ -0,0 +1,779 @@
+
+        Ok(())
+    }
+
+    #[tracing::instrument(level = Level::TRACE, skip(self), err)]
+    async fn handle_msg_from_task(
+        &mut self,
+        message: PortForwardMessage,
+    ) -> Result<(), PortForwardError> {
+        match message {
+            PortForwardMessage::Connect(remote_socket, oneshot) => {
+                let remote_address = SocketAddress::Ip(remote_socket);
+                self.oneshots.push_back(oneshot);
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Connect(
+                        LayerConnect { remote_address },
+                    )))
+                    .await
+                else {
+                    return Err(PortForwardError::AgentError(
+                        "failed to send ClientMessage to agent".into(),
+                    ));
+                };
+            }
+            PortForwardMessage::Send(connection_id, bytes) => {
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Write(
+                        LayerWrite {
+                            connection_id,
+                            bytes,
+                        },
+                    )))
+                    .await
+                else {
+                    return Err(PortForwardError::AgentError(
+                        "failed to send ClientMessage to agent".into(),
+                    ));
+                };
+            }
+            PortForwardMessage::Close(socket, connection_id) => {
+                // end of listener TCP stream in task
+                self.agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Close(
+                        LayerClose { connection_id },
+                    )))
+                    .await?;
+                self.tasks.remove(&socket);
+            }
+        }
+        Ok(())
+    }
+}
+
+struct LocalConnectionTask {
+    read_stream: ReaderStream<OwnedReadHalf>,
+    write: OwnedWriteHalf,
+    local_socket: SocketAddr,
+    remote_socket: SocketAddr,
+    task_internal_tx: Sender<PortForwardMessage>,
+    response_rx: Receiver<Vec<u8>>,
+}
+
+impl LocalConnectionTask {
+    pub fn new(
+        stream: TcpStream,
+        local_socket: SocketAddr,
+        remote_socket: SocketAddr,
+        task_internal_tx: Sender<PortForwardMessage>,
+        response_rx: Receiver<Vec<u8>>,
+    ) -> Self {
+        let (read, write) = stream.into_split();
+        let read_stream = ReaderStream::with_capacity(read, 64000);
+        Self {
+            remote_socket,
+            task_internal_tx,
+            response_rx,
+            read_stream,
+            write,
+            local_socket,
+        }
+    }
+
+    pub async fn run(&mut self) -> Result<(), PortForwardError> {
```

Comment:
This function exits early in multiple places. Make sure that main port-forwarder loop always knows about finished tasks (to do necessary clean-up).

Reply:
good point, i'm going to redesign run() slightly so that the loop returns a value which gets handled at the end of the function